### PR TITLE
Create local variables during PI modification

### LINKED ISSUE #9689: Modify a process instance by activating elements and setting local variables to a given scope
## Description

We want to pass variables when activating an element. The variables should be stored as local variables in the given scope.  Each activation instruction can contain variables and a different scope. 

The scope of the variables is selected by its element id. The element instance of the scope should be the same where the element is activated in. 

The variables should be set before activating the element. If the element is embedded in a subprocess then the variables should be set before activating the scope. The variables may be required to create event subscriptions of the element or of an upper flow scope.

For example:
image

* set the variables by writing a `created` event for each activate instruction
  * set the element instance key of the scope as the scope key of the variables

Out of scope:
* element id of scope doesn't exist

Blocked by 
- #9663

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

Using a callback create the variables during activation of of the element and flow scopes

## Related issues

closes #9689 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java [hunk 1]
```diff
@@ -268,85 +269,39 @@ public static class ProcessInstanceModificationClient {
...
-      if (variablesScopeId != null) {
-        variableInstruction.setElementId(variablesScopeId);
-      }
-      variableInstruction.setVariables(variables);
-      record.addActivateInstruction(
-          new ProcessInstanceModificationActivateInstruction()
-              .setElementId(elementId)
-              .addVariableInstruction(variableInstruction));
-      return this;
+    public ActivationInstructionBuilder activateElement(final String elementId) {
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java:299]: :+1: I like the new builder.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java [hunk 2]
```diff
@@ -64,15 +66,35 @@ public ElementActivationBehavior(
...
+   *
+   * @param processInstanceRecord the record of the process instance
+   * @param elementToActivate The element to activate
+   * @param createVariablesCallback Callback to create variables at a given scope
+   * @return The key of the activated element instance
+   */
+  public long activateElement(
+      final ProcessInstanceRecord processInstanceRecord,
+      final AbstractFlowElement elementToActivate,
+      final BiConsumer<DirectBuffer, Long> createVariablesCallback) {
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java:85]: :+1: The new callback fits well. The code is easy to understand.

**Review by saig0 — CHANGES_REQUESTED** [review]
@remcowesterhoud looks nice :+1: 

I have a few minor comments regarding the tests. Especially for event subprocesses, I'm not sure if this is already working as expected. Please have a look.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceVariablesTest.java [hunk 3]
```diff
@@ -197,4 +193,207 @@ public void shouldCreateVariablesBeforeElementScopes() {
...
+    assertThatVariableCreatedInScope(
+        processInstanceKey,
+        deployment.getValue().getProcessesMetadata().get(0).getProcessDefinitionKey(),
+        processInstanceKey,
+        "y",
+        "\"global\"");
+  }
+
+  @Test
+  public void shouldCreateLocalVariablesInNonExistingFlowscope() {
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceVariablesTest.java:289]: Wow, good catch, this was a bug 👍

**Review by saig0 — CHANGES_REQUESTED** [review]
@remcowesterhoud thank you for applying the hints :+1: 

I have one major point about the variable creation and a few minor suggestions. Please have another look. :cake:

**Review by saig0 — APPROVED** [review]
@remcowesterhoud super. Looks all good :+1:

**Comment by saig0** [comment]
bors merge

**Pull request merged**